src/Graph/GraphPrinter.php

```diff
@@ -87,197 +86,225 @@ class GraphPrinter extends SMWResultPrinter {
+						if ( $this->showGraphColor ) {
+							$graphInput .= "color=$color";
+						}
+						$graphInput .= ']';
+					}
 				}
+				$graphInput .= ';';
 			}
 		}
+		$graphInput .= "}";
+
+		// Calls graphvizParserHook function from MediaWiki GraphViz extension
+		$result = $GLOBALS['wgParser']->recursiveTagParse( "<graphviz>$graphInput</graphviz>" );
 
-		return implode( "\n", $segments );
+		// append legend
+		$result .= $this->getGraphLegend();
+
+		return $result;
 	}
 
 	/**
-	 * Returns the GV for a single SMWDataValue.
+	 * Creates the graph legend
 	 *
-	 * @since 1.5.4
+	 * @since 3.1
 	 *
-	 * @param SMWDataValue $object
-	 * @param $outputmode
-	 * @param boolean $isName Is this the name that should be used for the node?
-	 * @param string $name
-	 * @param string $labelName
+	 * @return string Html::rawElement
 	 *
-	 * @return string
 	 */
-	protected function getGVForDataValue( SMWDataValue $object, $outputmode, $isName, $name, $labelName ) {
-		$graphInput = '';
-		$nodeLabel = '';
-		$text = $object->getShortText( $outputmode );
-
-		if ( $this->m_graphLink ) {
-			$nodeLinkURL = "[[" . $text . "]]";
-		}
+	protected function getGraphLegend(){
+		if ( $this->showGraphLegend && $this->showGraphColor ) {
+			$itemsHtml = '';
+			$colorCount = 0;
+			$arraySize = count( $this->graphColors );
+
+			foreach ( $this->legendItem as $legendLabel ) {
+				if ( $colorCount >= $arraySize ) {
+					$colorCount = 0;
+				}
 
-		$text = $this->getWordWrappedText( $text, $this->m_wordWrapLimit );
+				$color = $this->graphColors[$colorCount];
+				$itemsHtml .= Html::rawElement( 'div', [ 'class' => 'graphlegenditem', 'style' => "color: $color" ],
+					"$color: $legendLabel" );
 
-		if ( $this->m_nodeLabel === self::NODELABEL_DISPLAYTITLE && $object instanceof SMWWikiPageValue ) {
-			$objectDisplayTitle = $object->getDisplayTitle();
-			if ( !empty( $objectDisplayTitle )) {
-				$nodeLabel = $this->getWordWrappedText( $objectDisplayTitle, $this->m_wordWrapLimit );
+				$colorCount ++;
 			}
-		}
 
-		if ( $this->m_graphLink ) {
-			if( $nodeLabel === '' ) {
-				$graphInput .= " \"$text\" [URL = \"$nodeLinkURL\"]; ";
-			} else {
-				$graphInput .= " \"$text\" [URL = \"$nodeLinkURL\", label = \"$nodeLabel\"]; ";
-			}
+			return Html::rawElement( 'div', [ 'class' => 'graphlegend' ], "$itemsHtml" );
 		}
+	}
 
-		if ( !$isName ) {
-			$graphInput .= $this->m_parentRelation ? " \"$text\" -> \"$name\" " : " \"$name\" -> \"$text\" ";
-
-			if ( $this->m_graphLabel || $this->m_graphColor ) {
-				$graphInput .= ' [';
-
-				if ( array_search( $labelName, $this->m_labelArray, true ) === false ) {
-					$this->m_labelArray[] = $labelName;
-				}
-
-				$color = $this->m_graphColors[array_search( $labelName, $this->m_labelArray, true )];
+	/**
+	 * Process a result row and create SRF\GraphNodes
+	 *
+	 * @since 3.1
+	 *
+	 * @param array $row
+	 *
+	 */
+	protected function processResultRow( array /* of SMWResultArray */ $row ) {
```

Comment:
This should probably be private. Try using private by default.

Generally speaking very few things are protected. In result printer classes you will need to have a bunch due to the inheritance design though.

If it is private it also should not have a @since tag